compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs

```diff
@@ -524,6 +554,86 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
+        &self,
+        cause: &ObligationCause<'tcx>,
+        param_env: ParamEnv<'tcx>,
+        deref_to: Ty<'tcx>,
+        deref_from: Ty<'tcx>,
+    ) -> Option<String> {
+        let tcx = self.infcx.tcx;
+
+        let deref_trait = tcx.lang_items().deref_trait()?;
+        let deref_mut_trait = tcx.lang_items().deref_mut_trait()?;
+        let deref_target = tcx.lang_items().deref_target()?;
+
+        // Count the number of references in deref_from, since we need
+        // that plus 1 to invoke Deref.
+        // This is also needed to ensure we aren't checking for Deref on a reference.
+        let mut deref_from_cur = deref_from;
+        // Extra required dereference to invoke Deref.
+        let mut num_refs = 1;
+        while let ty::Ref(_, inner_ty, _) = deref_from_cur.kind() {
+            deref_from_cur = *inner_ty;
+            num_refs += 1;
+        }
+
+        // <deref_from_cur as Deref>
+        let trait_ref = ty::TraitRef::new(tcx, deref_trait, [deref_from_cur]);
+        let obligation =
+            Obligation::new(tcx, cause.clone(), param_env, ty::Binder::dummy(trait_ref));
+        if !self.infcx.predicate_may_hold(&obligation) {
+            return None;
+        }
+
+        let ocx = ObligationCtxt::new(self.infcx);
+
+        // <deref_from_cur as Deref>::Target
+        let target = Ty::new_projection(tcx, deref_target, [deref_from_cur]);
+
+        let target_normalized = ocx.structurally_normalize(cause, param_env, target).ok()?;
+        let expected_normalized = ocx.structurally_normalize(cause, param_env, deref_to).ok()?;
+
+        // Determine whether target_normalized == expected_normalized with inference.
+        let types_eq = self.infcx.probe(|_| {
+            // FIXME: I don't know what DefineOpaqueTypes should be used here.
+            self.infcx
+                .at(cause, param_env)
+                .eq(DefineOpaqueTypes::No, expected_normalized, target_normalized)
+                .is_ok()
+        });
+
+        if !types_eq {
+            return None;
+        }
+
+        let deref_part = "*".repeat(num_refs);
+
+        // FIXME: Edge case: how should this handle double references?
+        //        Currently, they're dereferenced correctly, but only the
+        //        outer reference is added back after the dereferences.
+
+        // Check if we have DerefMut, as it's required to create a mut reference.
+        let trait_ref_deref_mut = ty::TraitRef::new(tcx, deref_mut_trait, [deref_from_cur]);
+        let obligation_deref_mut =
+            Obligation::new(tcx, cause.clone(), param_env, ty::Binder::dummy(trait_ref_deref_mut));
+        let has_deref_mut = self.infcx.predicate_may_hold(&obligation_deref_mut);
+
+        // Try to give a suggestion with the same type of reference as the original code.
+        // For example, if deref_from was a reference, then make the suggestion a reference as well.
+        let valid_mutability = deref_from
```

Comment:
I don't think we need to try to give mutable deref suggestions. I think borrowck will handle that once users try to apply these changes.